### PR TITLE
Add System Tests for Spectrum Viewer Features

### LINKED ISSUE #2320: Enhance System Tests for Spectrum Viewer
Review the existing system tests for the Spectrum Viewer to ensure comprehensive coverage. Address any identified gaps and update the tests as needed to improve their effectiveness.

### Desired Behaviour

The system tests for the Spectrum Viewer should comprehensively cover all key functionalities, ensuring reliable performance across all environments.

### Current Behaviour

The current system tests may have gaps, potentially missing critical functionalities or edge cases, risking undiscovered issues.

### PR DESCRIPTION
**Pull request by ashmeigh**
### Issue

Closes #2320

### Description

This pull request introduces unit tests for the Spectrum Viewer features. The changes focus on ensuring that all key functionalities related to Regions of Interest (ROIs) in the Spectrum Viewer are thoroughly tested.

ROIs Management: Tests for adding, removing, and renaming ROIs.

ROI Adjustments: Tests for adjusting the position and size of ROIs.

ROI Color: Tests for changing the color of ROIs.

Normalization Toggle: Tests for toggling normalization on and off.

Data Export: Tests for exporting data to CSV and RITS formats.

### Documentation

Release Notes: Added a note in the docs/release_notes about the new

### CODE REVIEW

File: mantidimaging/gui/test/gui_system_spectrum_test.py [hunk 1]
```diff
@@ -56,3 +57,71 @@ def test_add_roi(self) -> None:
...
+        self.assertNotIn(f'roi_{initial_roi_count - 1}', self.spectrum_window.spectrum_widget.roi_dict)
+
+    def test_change_roi_color(self) -> None:
+        QTest.mouseClick(self.spectrum_window.addBtn, Qt.MouseButton.LeftButton)
+        QTest.qWait(SHORT_DELAY)
+
+        new_color = (255, 0, 0, 255)
+        self.spectrum_window.spectrum_widget.change_roi_colour('roi_1', new_color)
+
+        self.assertEqual(self.spectrum_window.spectrum_widget.roi_dict['roi_1'].colour, new_color)
```
Comment (samtygier-stfc) [inline mantidimaging/gui/test/gui_system_spectrum_test.py:81]: This could step back a level, as `change_roi_colour()` is quite a deep place to call for system test.

I think it would work to mock `SpectrumROI.openColorDialog` and then call `onChangeColor`

File: mantidimaging/gui/test/gui_system_spectrum_test.py [hunk 2]
```diff
@@ -56,3 +57,71 @@ def test_add_roi(self) -> None:
...
+        QTest.mouseClick(self.spectrum_window.addBtn, Qt.MouseButton.LeftButton)
+        QTest.qWait(SHORT_DELAY)
+
+        roi_names = self.spectrum_window.presenter.get_roi_names()
+        roi_name = next(name for name in roi_names if name != 'all')
+
+        roi_widget = self.spectrum_window.spectrum_widget.roi_dict[roi_name]
+
+        roi_widget.setPos((roi_widget.pos().x() + 10, roi_widget.pos().y() + 10))
+        self.spectrum_window.presenter.handle_roi_moved()
```
Comment (samtygier-stfc) [inline mantidimaging/gui/test/gui_system_spectrum_test.py:104]: It might be possible to do this more realistically by moving the handle. Maybe something like
```
roi_widget.movePoint(0, (10, 20))
QTest.qWait(SHORT_DELAY)

updated_roi = self.spectrum_window.presenter.model.get_roi(roi_name)

self.assertEqual(updated_roi.left, 10)
self.assertEqual(updated_roi.top, 20)
```
That should trigger the signals for everything to get updated. You might need to check which handle number to put as the first arg to `movePoint`

File: mantidimaging/gui/test/gui_system_spectrum_test.py [hunk 3]
```diff
@@ -56,3 +57,71 @@ def test_add_roi(self) -> None:
...
+        assert self.spectrum_window.presenter.spectrum_mode == SpecType.SAMPLE
+
+    def test_export_csv(self) -> None:
+        with mock.patch("mantidimaging.gui.windows.spectrum_viewer.view.QFileDialog.getSaveFileName",
+                        return_value=("test_output.csv", "CSV Files (*.csv)")):
+            self.spectrum_window.presenter.handle_export_csv()
+
+            with open("test_output.csv") as file:
+                lines = file.readlines()
+                self.assertGreater(len(lines), 1)
```
Comment (samtygier-stfc) [inline mantidimaging/gui/test/gui_system_spectrum_test.py:127]: Lets drop this for now, as it would really need some file system mocking/faking which would be more complex.

File: mantidimaging/gui/test/gui_system_spectrum_test.py [hunk 4]
```diff
@@ -56,3 +57,71 @@ def test_add_roi(self) -> None:
...
+    def test_rename_roi(self) -> None:
+        QTest.mouseClick(self.spectrum_window.addBtn, Qt.MouseButton.LeftButton)
+        QTest.qWait(SHORT_DELAY)
+
+        old_name = 'roi_1'
+        new_name = 'roi_renamed'
+        self.spectrum_window.presenter.rename_roi(old_name, new_name)
+
+        self.assertNotIn(old_name, self.spectrum_window.spectrum_widget.roi_dict)
+        self.assertIn(new_name, self.spectrum_window.spectrum_widget.roi_dict)
```
Comment (samtygier-stfc) [inline mantidimaging/gui/test/gui_system_spectrum_test.py:92]: Would be good here to step back and trigger it by modifying the table. It might work to use
`self.spectrum_window.presenter.roi_table_model.set_element()` to change the name, and then call `self.spectrum_window.presenter.roi_table_model.dataChanged.emit()`, or something similar.

**Review by samtygier-stfc — CHANGES_REQUESTED** [review]
These are useful, but a few a of them need to step back a level so that they are exercising the gui more than what could be done in unit tests.

File: mantidimaging/gui/test/gui_system_spectrum_test.py [hunk 5]
```diff
@@ -56,3 +57,78 @@ def test_add_roi(self) -> None:
...
+
+        roi_name = 'roi_1'
+        new_color = (255, 0, 0, 255)
+
+        sensible_roi = SensibleROI(left=0, top=0, right=10, bottom=10)
+        self.spectrum_window.spectrum_widget.add_roi(sensible_roi, roi_name)
+
+        with mock.patch('mantidimaging.gui.windows.spectrum_viewer.spectrum_widget.SpectrumROI.openColorDialog',
+                        return_value=new_color):
+            self.spectrum_window.spectrum_widget.roiColorChangeRequested.emit(roi_name, new_color)
```
Comment (samtygier-stfc) [inline mantidimaging/gui/test/gui_system_spectrum_test.py:86]: This can take another step back and call `onChangeColor`. It needed a tweak to the mocking so that the return value was a QColor.
```python
        with mock.patch('mantidimaging.gui.windows.spectrum_viewer.spectrum_widget.SpectrumROI.openColorDialog',
                        return_value=QColor(*new_color)):
            self.spectrum_window.spectrum_widget.roi_dict[roi_name].onChangeColor()
```

You could also use `mock.patch.object` because we have a reference to the thing we are patching.
```python
        spec_roi = self.spectrum_window.spectrum_widget.roi_dict[roi_name]
        with mock.patch.object(spec_roi, 'openColorDialog', return_value=QColor(*new_color)):
            spec_roi.onChangeColor()
```

File: mantidimaging/gui/test/gui_system_spectrum_test.py [hunk 6]
```diff
@@ -56,3 +57,78 @@ def test_add_roi(self) -> None:
...
+        table_model = self.spectrum_window.roi_table_model
+        row = table_model.roi_names().index(old_name)
+
+        table_model.set_element(row, 0, new_name)
+        table_model.dataChanged.emit(table_model.index(row, 0), table_model.index(row, 0))
+
+        self.assertNotIn(old_name, self.spectrum_window.spectrum_widget.roi_dict)
+        self.assertIn(new_name, self.spectrum_window.spectrum_widget.roi_dict)
+
+    def test_adjust_roi(self):
```
Comment (samtygier-stfc) [inline mantidimaging/gui/test/gui_system_spectrum_test.py:106]: Did you try with `movePoint`?

**Review by samtygier-stfc — APPROVED** [review]
Looks good.

I squashed the commits to remove the merges and some churn from adding and removing things.
